File: kb/family_kb.pl
% -------- FACTS (15) --------
parent(john, mary).
parent(mary, sue).
parent(john, alex).
parent(alex, david).
parent(david, lily).

male(john).
male(alex).
male(david).

female(mary).
female(sue).
female(lily).
female(anna).

married(john, anna).
married(anna, john).

% -------- RULES (8) --------
ancestor(X, Y) :- parent(X, Y).
ancestor(X, Y) :- parent(X, Z), ancestor(Z, Y).

grandparent(X, Y) :- parent(X, Z), parent(Z, Y).

sibling(X, Y) :- parent(Z, X), parent(Z, Y), X \= Y.

mother(X, Y) :- parent(X, Y), female(X).
father(X, Y) :- parent(X, Y), male(X).

spouse(X, Y) :- married(X, Y).

brother(X, Y) :- sibling(X, Y), male(X).

File: kb/family_kb.py
from fol_bc import Rule, KnowledgeBase

def build_kb():
    rules = [
        # FACTS
        Rule(("parent","john","mary"), ()),
        Rule(("parent","mary","sue"), ()),
        Rule(("parent","john","alex"), ()),
        Rule(("parent","alex","david"), ()),
        Rule(("parent","david","lily"), ()),

        Rule(("male","john"), ()),
        Rule(("male","alex"), ()),
        Rule(("male","david"), ()),

        Rule(("female","mary"), ()),
        Rule(("female","sue"), ()),
        Rule(("female","lily"), ()),
        Rule(("female","anna"), ()),

        Rule(("married","john","anna"), ()),
        Rule(("married","anna","john"), ()),

        # RULES
        Rule(("ancestor","?x","?y"), (("parent","?x","?y"),)),
        Rule(("ancestor","?x","?y"), (("parent","?x","?z"), ("ancestor","?z","?y"))),

        Rule(("grandparent","?x","?y"), (("parent","?x","?z"), ("parent","?z","?y"))),

        Rule(("sibling","?x","?y"), (("parent","?z","?x"), ("parent","?z","?y"))),  # note: no X != Y check here unless you implemented it

        Rule(("mother","?x","?y"), (("parent","?x","?y"), ("female","?x"))),
        Rule(("father","?x","?y"), (("parent","?x","?y"), ("male","?x"))),

        Rule(("spouse","?x","?y"), (("married","?x","?y"),)),

        Rule(("brother","?x","?y"), (("sibling","?x","?y"), ("male","?x"))),
    ]
    return KnowledgeBase(rules)
